File: joe/typescript/rps.ts
import { createInterface } from 'readline';
import { once } from 'events';

const enum Choice {
	Rock = 'rock',
	Paper = 'paper',
	Scissors = 'scissors',
}

const GREEN_ANSI_START = '\u001b[92m';
const ANSI_END = '\u001b[0m';

const winCases = {
	[Choice.Rock]: Choice.Scissors,
	[Choice.Scissors]: Choice.Rock,
	[Choice.Paper]: Choice.Rock,
};
const choices = [Choice.Rock, Choice.Paper, Choice.Scissors];

// Small utility to prompt for input. Wrapped in an IIFE to prevent polluting scope with `scanner`.
const ask = (() => {
	const scanner = createInterface({ input: process.stdin, output: process.stdout });
	return async (question: string) => {
		// We don't want a newline
		process.stdout.write(`${GREEN_ANSI_START}${question}${ANSI_END}`);
		const [input]: string[] = await once(scanner, 'line');
		return input.trim().toLowerCase();
	};
})();

async function main() {
	for (;;) {
		const userChoice = await ask('Rock, paper, scissors? ');

		switch (userChoice) {
			case Choice.Rock:
			case Choice.Scissors:
			case Choice.Paper:
				// Okay
				break;

			case '.quit':
			case 'q':
			case 'stop':
			case 'exit':
				console.log('Fun playing with you, shutting down...');
				process.exit(0);

			default:
				console.log("That wasn't a valid choice, please try again.");
				continue;
		}

		const botChoice = choices[Math.floor(Math.random() * choices.length)];
		const losingChoice = winCases[botChoice];
		if (userChoice === losingChoice) {
			console.log(
				`You got absolutely destroyed by the bigbrain computer, as I chose ${botChoice} versus your ${userChoice}. Wanna play again?`,
			);
		} else if (userChoice === botChoice) {
			console.log("Aw man, it was a tie... Let's play again.");
		} else {
			console.log(`You beat me, human. I chose ${botChoice} but you chose ${userChoice}... Now let's play again.`);
		}
	}
}

main();

File: joe/golang/rps.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"
)

const (
	ChoiceRock     = "rock"
	ChoicePaper    = "paper"
	ChoiceScissors = "scissors"

	GreenAnsiStart = "\033[92m"
	AnsiEnd        = "\033[0m"
)

var (
	winCases = map[string]string{ChoiceRock: ChoiceScissors, ChoiceScissors: ChoicePaper, ChoicePaper: ChoiceRock}
	choices  = [3]string{ChoiceRock, ChoicePaper, ChoiceScissors}
)

func main() {
	rand.Seed(time.Now().UnixNano())

	for {
		choice := input("Rock, paper, scissors? ")

		switch choice {
		case ChoiceRock, ChoicePaper, ChoiceScissors:
			// ok
		case ".quit", "q", "stop", "exit":
			fmt.Println("Fun playing with you, shutting down...")
			os.Exit(0)
		default:
			fmt.Println("That wasn't a valid choice, please try again.")
			continue
		}

		botChoice := choices[rand.Intn(3)]

		loseCase := winCases[botChoice]

		switch {
		case choice == loseCase:
			fmt.Printf("You got absolutely destroyed by the bigbrain computer, as I chose %s versus your %s. Wanna play again?\n", botChoice, choice)
		case choice == botChoice:
			fmt.Println("Aw man, it was a tie... Let's play again.")
		default:
			fmt.Printf("You beat me, human. I chose %s but you chose %s... Now let's play again.\n", botChoice, choice)
		}
	}
}

func input(prompt string) string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(prompt)
	text, _ := reader.ReadString('\n')

	return strings.ToLower(strings.TrimSpace(text))
}
